google/cloud/pubsub_v1/subscriber/_protocol/leaser.py

```diff
@@ -128,7 +128,9 @@ def maintain_leases(self):
             # Determine the appropriate duration for the lease. This is
             # based off of how long previous messages have taken to ack, with
             # a sensible default and within the ranges allowed by Pub/Sub.
-            deadline = self._manager.ack_deadline
+            # Also update the deadline currently used if enough new ACK data has been
+            # gathered since the last deadline update.
+            deadline = self._manager._obtain_ack_deadline(maybe_update=True)
```

Comment:
What's the rationale for updating the ack deadline here?

Why not in the dispatcher ack method after updating the histogram?

Reply:
If an ACK deadline is updated while the leaser is sleeping (let's assume it gets shorter), newly arrived messages will use that updated deadline. This could be problematic if processing of these messages gets delayed - the (new) deadline would expire, because the leaser would still be sleeping, as it went to sleep using the older (longer) deadline.

It's actually the very issue this PR fixes.

Reply:
IOW, we can't _decrease_ the deadline while the leaser is sleeping, because then it wouldn't wake in time to renew, as it would be sleeping based on the old longer deadline.

Reply:
Correct, you summarized it well.